File: Module Service/src/main/java/com/abetiou/moduleservice/Services/ModuleService.java
package com.abetiou.moduleservice.Services;

import com.abetiou.moduleservice.DTO.ModuleWithStudentCount;
import com.abetiou.moduleservice.DTO.StudentDto;
import com.abetiou.moduleservice.Entities.CourseModule;
import com.abetiou.moduleservice.Repository.ModuleRepository;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class ModuleService {

    private final ModuleRepository moduleRepository;
    private final RestTemplate restTemplate;


    public ModuleService(ModuleRepository moduleRepository,RestTemplate restTemplate) {
        this.moduleRepository = moduleRepository;
        this.restTemplate = restTemplate;
    }

    // Méthode pour créer un module et l'affecter à un professeur
    public CourseModule createModule(String moduleName, int nombreHeures,String semestre, Long profId) {
        CourseModule courseModule = new CourseModule(null, moduleName, nombreHeures,semestre, profId, null);
        return moduleRepository.save(courseModule);
    }


    //AFFECTATION STUDNET <==> MODULE
    public CourseModule assignStudentsToModule(Long moduleId, String niveau) {
        // URL pour récupérer les étudiants par niveau
        String url = "http://localhost:8082/students/allStudentsWithNiveau/" + niveau;

        // URL pour assigner le module aux étudiants
        String url2 = "http://localhost:8082/students/assignModuleToStudent";

        // Créer un RestTemplate
        RestTemplate restTemplate = new RestTemplate();

        // 1. Appel pour récupérer la liste des étudiants par niveau
        ResponseEntity<List<StudentDto>> response;
        try {
            response = restTemplate.exchange(
                    url, HttpMethod.GET, null, new ParameterizedTypeReference<List<StudentDto>>() {});
        } catch (Exception e) {
            throw new RuntimeException("Failed to fetch students from service Student: " + e.getMessage());
        }

        // Vérification de la réponse
        if (!response.getStatusCode().is2xxSuccessful() || response.getBody() == null) {
            throw new RuntimeException("Failed to fetch students or empty response");
        }

        // Extraire les IDs des étudiants
        Set<Long> studentIds = response.getBody().stream()
                .map(StudentDto::getId)
                .collect(Collectors.toSet());

        // 2. Appel pour assigner le module aux étudiants récupérés
        // Préparer les headers de la requête
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        // Créer les paramètres de la requête
        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        params.add("moduleId", moduleId.toString());
        params.add("niveau", niveau);

        // Préparer la requête
        HttpEntity<MultiValueMap<String, String>> requestEntity = new HttpEntity<>(params, headers);

        try {
            restTemplate.exchange(url2, HttpMethod.POST, requestEntity, Void.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to assign module to students: " + e.getMessage());
        }

        // 3. Récupérer le module existant
        CourseModule module = moduleRepository.findById(moduleId)
                .orElseThrow(() -> new RuntimeException("Module not found with ID: " + moduleId));

        // Ajouter les IDs des étudiants au module
        module.setStudentIds(studentIds);

        // 4. Sauvegarder le module avec les étudiants assignés
        return moduleRepository.save(module);
    }


    //trouver module par id
    public CourseModule findModuleById(Long moduleId) {
        return moduleRepository.findById(moduleId)
                .orElseThrow(() -> new RuntimeException("Module not found with ID: " + moduleId));
    }


    public List<CourseModule> courseModuleList(){
        return moduleRepository.findAll();
    }


    public List<CourseModule> getModulesByProfId(Long profId) {
        return moduleRepository.findByProfId(profId);
    }

    public List<ModuleWithStudentCount> getModulesWithStudentCount() {
        // Ensure `findAll()` is called on the instance
        return moduleRepository.findAll().stream()
                .map(module -> new ModuleWithStudentCount(
                        module.getId(),
                        module.getModuleName(),
                        module.getNombreHeures(),
                        module.getSemestre(),
                        module.getProfId(),
                        module.getStudentIds().size()
                ))
                .collect(Collectors.toList());
    }


}

File: Frontend/src/app/services/module.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CourseModule } from '../models/module.model';
import { StudentModule } from '../models/student-module.model';

@Injectable({
  providedIn: 'root'
})
export class ModuleService {
  private baseUrl = 'http://localhost:8084/modules';  // L'URL de base pour les modules

  constructor(private http: HttpClient) {}

  // Méthode pour récupérer les modules par ID du professeur
  getModulesByProfId(profId: number): Observable<CourseModule[]> {
    return this.http.get<CourseModule[]>(`${this.baseUrl}/prof/${profId}`);
  }


  createModule(moduleData: { moduleName: string; nombreHeures: number; semestre: string; profId: number }): Observable<CourseModule> {
    const params = new URLSearchParams();
    params.set('moduleName', moduleData.moduleName);
    params.set('nombreHeures', moduleData.nombreHeures.toString());
    params.set('semestre', moduleData.semestre);
    params.set('profId', moduleData.profId.toString());

    return this.http.post<CourseModule>(`${this.baseUrl}/create?${params.toString()}`, {});
  }


  getAllModules(): Observable<CourseModule[]> {
    return this.http.get<CourseModule[]>(`${this.baseUrl}/allmodules`);
  }

   // Affecter des étudiants à un module
   // Affecter des étudiants à un module
   assignStudentsToModule(moduleId: number, niveau: string): Observable<CourseModule> {
    return this.http.put<CourseModule>(`${this.baseUrl}/${moduleId}/assign-students`, null, {
      params: { niveau }
    });
  }

  getStudentsByModule(moduleId: number): Observable<StudentModule[]> {
    return this.http.get<StudentModule[]>(`http://localhost:8082/students/by-module/${moduleId}`);
  }

  removeModuleFromStudent(studentId: number, moduleId: number): Observable<void> {
    return this.http.delete<void>(`http://localhost:8082/students/${studentId}/modules/${moduleId}`);
  }


  deleteModuleByProf(profId: number, moduleId: number) {
    const url = `http://localhost:8084/modules/delete?moduleId=${moduleId}&profId=${profId}`;
    return this.http.delete<any>(url);
  }





}
